packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -91,139 +121,548 @@ const propTypes = {
     }
 
-    const format = intl.formatMessage({ id: 'Terra.datePicker.dateFormat' });
+    setDate(inputValue, type);
+  };
+
+  const handleDayChange = (event) => {
+    let inputValue = event.target.value;
+    if (!DateUtil.validDateInput(inputValue)) {
+      return;
+    }
+
+    // Ignore the entry if the value did not change or it is invalid.
+    // When 'Predictive text' is enabled on Android the maxLength attribute on the input is ignored so we have to
+    // check the length of inputValue to make sure that it is less then 2.
+    if (inputValue === date.day || inputValue.length > 2 || Number(inputValue) > 31 || inputValue === '00') {
+      return;
+    }
+
+    // If the change made was not a deletion of a digit, then prepend '0'
+    // when the input value is a single digit value between 4 and 9
+    if (inputValue.length >= date.day.length) {
+      const digitsToPrependZero = ['4', '5', '6', '7', '8', '9'];
 
-    return (
-      <div className={cx(theme.className)}>
-        <div className={cx('custom-input')}>
+      if (digitsToPrependZero.indexOf(inputValue) > -1) {
+        inputValue = `0${inputValue}`;
+      }
+    }
+
+    handleDateChange(event, inputValue, DateUtil.inputType.DAY);
+  };
+
+  const handleMonthChange = (event) => {
+    let inputValue = event.target.value;
+    if (!DateUtil.validDateInput(inputValue)) {
+      return;
+    }
+
+    // Ignore the entry if the value did not change or it is invalid.
+    // When 'Predictive text' is enabled on Android the maxLength attribute on the input is ignored so we have to
+    // check the length of inputValue to make sure that it is less then 2.
+    if (inputValue === date.month || inputValue.length > 2 || Number(inputValue) > 12 || inputValue === '00') {
+      return;
+    }
+
+    // If the change made was not a deletion of a digit, then prepend '0'
+    // when the input value is a single digit value between 2 and 9
+    if (inputValue.length >= date.month.length) {
+      const digitsToPrependZero = ['2', '3', '4', '5', '6', '7', '8', '9'];
+
+      if (digitsToPrependZero.indexOf(inputValue) > -1) {
+        inputValue = `0${inputValue}`;
+      }
+    }
+
+    handleDateChange(event, inputValue, DateUtil.inputType.MONTH);
+  };
+
+  const handleYearChange = (event) => {
+    const inputValue = event.target.value;
+    if (!DateUtil.validDateInput(inputValue)) {
+      return;
+    }
+
+    // Ignore the entry if the value did not change or it is invalid.
+    // When 'Predictive text' is enabled on Android the maxLength attribute on the input is ignored so we have to
+    // check the length of inputValue to make sure that it is less then 4.
+    if (inputValue === date.year || inputValue.length > 4) {
+      return;
+    }
+
+    // Ignore the entry if the year value is not between 1900 and 2100
+    if (inputValue.length === 4 && (Number(inputValue) < 1900 || Number(inputValue) > 2100)) {
+      return;
+    }
```

Comment:
The behavior in the Year input is different from the month, day, hour, and minute inputs. Yes, the current logic  protects against typing/pasting in an out-of-range year, but is different than the other fields in that it lets you type three digits before figuring out the number is out of range, and never has any autocomplete which helps resolve it to a valid value.

Let's update the year behavior to be:
- User can type 0 - 99, but if the first two digits do not match 19, 20, or 21, the additional digits are prevented from being typed. 
- If the user types 19 or 20, they can continue typing additional digits, 0 - 99.
- If the user types 21, they can only type 0 and 0.
- When focus is changed away from input, if the value is one digit (0-9) --> the input will prepend '200', making the year two thousand and {typed number}. (similar to the month and day prepending a zero)
- When focus is changed away from input, if the value is two digits (00-99) --> the input will prepend '20', making the year two thousand and {typed number}. (similar to the month and day prepending a zero)
- When focus is changed away from input, if the value is three digits (190-210) --> the input will append '0', making the year match a rounded decade. e.g. 190 -> 1900, 193 -> 1930, 200 -> 2000, 202 -> 2020, 209 -> 2090, 210 -> 2100.
- Continue to protect against pasting a value that is outside of the 1900-2100 mix/max range.

Reply:
Added the appropriate filter to the year input : https://github.com/cerner/terra-framework/pull/1208/commits/1d83d4d33f91233ca936e2e9c19f4428246a32d8

Reply:
When you enter `0` for year, hour, minute, or second and focus is changed, it would prepend/default to the intended value. However, if you enter `0` for day or month, the value stays at 0 and this behavior is inconsistent with the other inputs. Should it default to `01`?

Reply:
I agree that it did seem inconsistent, so i have changed it here https://github.com/cerner/terra-framework/pull/1208/commits/05543dce4fa33f2dfe8995457f558ae652b48e82 and am now defaulting to `01` is just `0` is entered.
@neilpfeiffer, can you verify if this behaviour is appropriate?

Reply:
@neilpfeiffer, regarding the 4th and 5th point above for prepending the year input value when the value is one or two digits and the focus is changed away from the input. instead of adding a hardcoded value of  '200' or '20'. should we consider restricting user entry as mentioned below ?  as the allowed year input range is between 1900-2100 and also user always tries to enter from the beginning of the year Left-Right like for 2021. user usually types in this order 2 -> 0 -> 2 -> 1.

1. For the first digit entry of the year input to be always restricted to either 1 or 2.
2. Second digit entry to be restricted to 9 if the first digit is 1.  0 or 1 in case first digit is 2 
3. Third digit could be anything between (0-9) if the first two digits are '19'  or '20' . if the first two digits is '21' then the third digit entry should be restricted to zero
4. Fourth digit could be anything between (0-9). but if first three digits are '210' then user entry should be restricted to zero . 
5. As mentioned above if user enters the first digit and it is either 1 or 2. when the focus is moved out in this case append 900 if the first input is 1 and append 100 if the first digit is 2.
6. As mentioned above if user enters the first two digits and it is either 19, 20 or 21. when the focus is moved out, in this case always append '00'.
7. For the fourth digit always append zero if the focus is moved out after entering the first three digits for the year input.

I understand this might make our logic more complex and even we need to do extra validation while deleting the year input digits one by one and in between once it is entered.

Reply:
My concern with this proposed behavior is that it would be inconsistent with how time-input currently behaves...unless we also want to change how time-input works. Take the hour input as an example, users are allowed to enter 0-9 as the first digit even though the first digit can only be 0 or 1 for 12-hour notation or 0,1, or 2 for 24-hour notation.